Rubberduck.Parsing/Rewriter/AttributesRewriteSession.cs

```diff
@@ -0,0 +1,41 @@
+﻿using System;
+using Rubberduck.Parsing.VBA;
+using Rubberduck.VBEditor;
+
+namespace Rubberduck.Parsing.Rewriter
+{
+    public class AttributesRewriteSession : RewriteSessionBase
+    {
+        private readonly RubberduckParserState _state;
+
+        public AttributesRewriteSession(RubberduckParserState state, IRewriterProvider rewriterProvider,
+            Func<IRewriteSession, bool> rewritingAllowed)
+            : base(rewriterProvider, rewritingAllowed)
+        {
+            _state = state;
+        }
+
+        protected override IExecutableModuleRewriter ModuleRewriter(QualifiedModuleName module)
+        {
+            return RewriterProvider.CodePaneModuleRewriter(module);
+        }
+
+        protected override void RewriteInternal()
+        {
+            //The suspension ensures that only one parse gets executed instead of two for each rewritten module.
+            var result = _state.OnSuspendParser(this, new[] {ParserState.Ready}, RewriteAllRewriters);
+            if(result != SuspensionResult.Completed)
+            {
+                Logger.Warn($"Rewriting attribute modules did not succeed. suspension result = {result}");
+            }
+        }
+
+        private void RewriteAllRewriters()
+        {
+            foreach (var rewriter in CheckedOutModuleRewriters.Values)
+            {
+                rewriter.Rewrite();
```

Comment:
This does not really have to do anything with ANTLR. All we get from the rewriters is `GetText`. Then we swap the module code in the code panes or the files. The call to `GetText` actually should never fail. So any problem arising would be connected to us swapping the code. 

Unfortunately, from the top of my head I do not see a good way to implement a role-back mechanism for swapping the code in the modules.

Reply:
Hitting Ctrl+Z as needed should revert all changes in most cases; one of the reasons for using rewriters was to preserve as much of the undo stack as we could. Of course if many modules (more than half the undo stack's capacity) are in the session, then we're out of luck. One solution could be to cache each module's contents prior to a rewrite operation, and implement our own rollback.

Reply:
I like this PR.  I've been working on a `MoveMember` refactoring and I'm thinking the following will come up for `MoveMember` or any other refactoring that includes a preview: (First, I am assuming that `rewriter.GetText()` would be called to load the preview)...if the user is provided various options that change the rewrite, is there (or will there be) a way to reset the associated `rewriter`?  e.g., `RenameRefactoring`: upon preview, the `Replace` command will have already been called on the `rewriter` so `rewriter.GetText()` generates the preview...if the user is allowed to change the name during preview, the earlier `Replace` command would need to be 'removed' and re-applied with the new name.  With this PR, would/could the roll-back be accomplished by opening a new `RewriteSession` and re-applying the changes?

Reply:
Opening a new rewrite session would provide a new session in which no rewriter is checked out. Checking out one, will provide a new one, because the session has not checked it out yet. So yes, getting a new session resets the rewriter.

This brings to mind that adding a `ResetRewriter` method might be useful, which resets a specific rewriter, should only that be required to be reset. (It would simply remove the rewriter from the dictionary of checked out rewriters preventing it from being applied and making the next checkout of the module provide a new one.)